### ISSUE TITLE
Error drawing visual vispy visuals mesh.MeshVisual and vispy visuals tube.TubeVisual

### ISSUE THREAD

**Issue by kunnalparihar** [vispy/vispy#1667]
I have upgraded vispy today to 0.6.0 and now I am not able to run this example
I am getting empty canvas screen and widgets box window.
https://github.com/vispy/vispy/blob/master/examples/demo/visuals/wiggly_bar.py

I got this error

WARNING: Error drawing visual <vispy.visuals.mesh.MeshVisual object at 0x7ff6540694a8>
WARNING:vispy:Error drawing visual <vispy.visuals.mesh.MeshVisual object at 0x7ff6540694a8>
WARNING: Traceback (most recent call last):
  File "wiggly_vispy.py", line 855, in <module>
    main()
  File "wiggly_vispy.py", line 851, in main
    win.show()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/app/backends/_qt.py", line 818, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "wiggly_vispy.py", line 344, in on_draw
    vis.draw()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/visual.py", line 595, in draw
    v.draw()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/mesh.py", line 519, in draw
    Visual.draw(self, *args, **kwds)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/visual.py", line 443, in draw
    self._vshare.index_buffer)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/shaders/program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/program.py", line 533, in draw
    canvas.context.flush_commands()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 572, in flush
    self._shared.flush(parser)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 494, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 819, in parse
    self._parse(command)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 779, in _parse
    ob.draw(*args)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 1322, in draw
    gl.check_error('Check before draw')
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/gl/__init__.py", line 207, in check_error
    raise err
RuntimeError: OpenGL got errors (Check before draw): GL_INVALID_VALUE
WARNING:vispy:Traceback (most recent call last):
  File "wiggly_vispy.py", line 855, in <module>
    main()
  File "wiggly_vispy.py", line 851, in main
    win.show()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/app/backends/_qt.py", line 818, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "wiggly_vispy.py", line 344, in on_draw
    vis.draw()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/visual.py", line 595, in draw
    v.draw()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/mesh.py", line 519, in draw
    Visual.draw(self, *args, **kwds)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/visual.py", line 443, in draw
    self._vshare.index_buffer)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/visuals/shaders/program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/program.py", line 533, in draw
    canvas.context.flush_commands()
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 572, in flush
    self._shared.flush(parser)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 494, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 819, in parse
    self._parse(command)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 779, in _parse
    ob.draw(*args)
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/glir.py", line 1322, in draw
    gl.check_error('Check before draw')
  File "/home/kunnal/.local/lib/python3.5/site-packages/vispy/gloo/gl/__init__.py", line 207, in check_error
    raise err
RuntimeError: OpenGL got errors (Check before draw): GL_INVALID_VALUE
ERROR: Invoking <bound method WigglyBar.on_draw of <WigglyBar (PyQt5) at 0x7ff65455d9e8>> for DrawEvent
ERROR:vispy:Invoking <bound method WigglyBar.on_draw of <WigglyBar (PyQt5) at 0x7ff65455d9e8>> for DrawEvent
WARNING: Error drawing visual <vispy.visuals.mesh.MeshVisual object at 0x7ff6540694a8>
WARNING:vispy:Error drawing visual <vispy.visuals.mesh.MeshVisual object at 0x7ff6540694a8>
ERROR: Invoking <bound method WigglyBar.on_draw of <WigglyBar (PyQt5) at 0x7ff65455d9e8>> repeat 2
ERROR:vispy:Invoking <bound method WigglyBar.on_draw of <WigglyBar (PyQt5) at 0x7ff65455d9e8>> repeat 2
WARNING: Error drawing visual <vispy.visuals.tube.TubeVisual object at 0x7ff654048dd8>

Thanks for help

**Comment by kmuehlbauer** [comment]
@kunnalparihar Thanks for the report. I got similar traceback using python 3.7. Unfortunately this demo isn't tested within the CI Framework. 

@djhoese Seems that also other Apps which use PyQt are broken (eg https://github.com/vispy/vispy/blob/master/examples/demo/gloo/primitive_mesh_viewer_qt.py)

**Comment by djhoese** [comment]
Hm, I get the same results. For the mesh viewer example:

```
$ python examples/demo/gloo/primitive_mesh_viewer_qt.py
WARNING: Traceback (most recent call last):
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 818, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 428, in __call__
    raise RuntimeError('EventEmitter loop detected!')
RuntimeError: EventEmitter loop detected!
Abort trap: 6
```

**Comment by asnt** [comment]
For info, on my side, same error as the first one reported above for:
```
examples/demo/visuals/wiggly_bar.py
examples/demo/gloo/two_qt_widgets.py
```
However, `examples/demo/gloo/primitive_mesh_viewer_qt.py` seems to work without error.

Linux/Python 3.7

**Comment by kunnalparihar** [comment]
Yes, In my case also examples/demo/gloo/primitive_mesh_viewer_qt.py seems to work without error. Actual problem comes when I am using vispy.visuals with PyQt5 in latest version (0.6.0) of vispy.

**Comment by kmuehlbauer** [comment]
@asnt @kunnalparihar @djhoese the output of `vispy.sys_info` would be interesting. And, we should try to narrow it down by simplifying the examples.

**Comment by kunnalparihar** [comment]
@asnt @djhoese 
the output when I run vispy.sys_info. I have made it short. Do you need full output ?

"Platform: Linux-4.15.0-54-generic-x86_64-with-Ubuntu-16.04-xenial
Python: 3.5.2 (default, Nov 12 2018, 13:43:14)  [GCC 5.4.0 20160609]
NumPy:    1.16.2
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.12.2', '5.12.3')
pyside:   None
pyside2:  None
pyglet:   None
glfw:     glfw (3, 1, 2)
sdl2:     None
wx:       None
egl:      EGL 1.4 (DRI2) Mesa Project: OpenGL OpenGL_ES
osmesa:   OSMesa
_test:    None
GL version:  '3.0 Mesa 18.0.5'
MAX_TEXTURE_SIZE: 16384
Extensions: 'GL_3DFX_texture_compression_FXT1 GL_AMD_conservative_depth......

**Comment by kmuehlbauer** [comment]
@kunnalparihar No, that will be sufficient. I doubt, that the Extensions have anything to do with it.

**Comment by kmuehlbauer** [comment]
My current setup with vispy 0.6.0 from conda-forge.
```
Platform: Linux-4.12.14-lp150.12.64-default-x86_64-with-glibc2.10
Python:   3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)  [GCC 7.3.0]
NumPy:    1.16.4
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.9.2', '5.9.7')
pyside:   None
pyside2:  None
pyglet:   None
glfw:     None
sdl2:     None
wx:       None
egl:      None
osmesa:   OSMesa
_test:    None

GL version:  '4.6.0 NVIDIA 390.87'
MAX_TEXTURE_SIZE: 16384
```

**Comment by kmuehlbauer** [comment]
@djhoese For the meshviewer demo (https://github.com/vispy/vispy/blob/master/examples/demo/gloo/primitive_mesh_viewer_qt.py) it seems that vispy app event handling collides with pyqt event handling. If I change the canvas to scene.SceneCanvas then there is at least some output and the different meshes can be altered. I do not have the time to dig deeper, but something is fishy with the eventhandling.

**Comment by asnt** [comment]
On my side:
```
Platform: Linux-5.1.15-arch1-1-ARCH-x86_64-with-arch
Python:   3.7.3 (default, Jun 24 2019, 04:54:02)  [GCC 9.1.0]
NumPy:    1.16.4
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.12.3', '5.13.0')
pyside:   None
pyside2:  None
pyglet:   pyglet 1.4.0b1
glfw:     glfw (3, 3, 0)
sdl2:     None
wx:       None
egl:      EGL 1.4 Mesa Project: OpenGL OpenGL_ES
osmesa:   OSMesa
_test:    None

GL version:  '3.0 Mesa 19.1.1'
MAX_TEXTURE_SIZE: 16384
```

**Comment by djhoese** [comment]
Ok major improvement on the mesh viewer: commenting out the fps measurement makes it work 100%. With it, on my newest python 3.7 conda-forge based environment with PyQt5 5.9, I get a lot of issues where the GL canvas won't draw and most times even the Qt widgets won't draw. This makes me think the FPS measurement is corrupting the GL context in some way or stopping Qt from drawing things (since all Widgets in Qt5 are GL based iirc). Will investigate further...

**Comment by djhoese** [comment]
@kmuehlbauer I messaged you on gitter to see if you could look at the mesh viewer example for me and see if you see the same results. If you comment out the `showMessage` call in the `show_fps` method you should see that everything works as expected. If you print the FPS that's fine. If you use `showMessage` though it causes a recursive repaint event where `showMessage` starts a draw, which calls the measure fps callback, which results in `showMessage`, and so on. If you give it a message that doesn't change (a static string), Qt will "debounce" it and it works fine.

I tried to see if it was something specific to the QStatusBar that comes with the main widget by adding a QLabel to the widgets on the left of the Window and set the text on it instead of the status bar. This "works" but it causes a weird drawing effect where the mesh goes back and forth from the right size in the center of the canvas to a smaller size in the lower left. This may be a problem related to me having a HiDPI display (which makes me a little scared).

Basically, I can't tell if this is a bug in PyQt or bad practice by this example. I *think* it makes sense that `showMessage` repaints *everything*, especially given the little I understand about PyQt5 using OpenGL for all widgets. I'm wondering if we maybe put the measure fps callback *before* the regular draw callbacks (not sure we can do that with the way it disconnects/connects repeatedly) if that will fix things. I haven't tested that.

**Comment by djhoese** [comment]
Oh and on the wiggly bar: so far if I comment out `on_draw` then it at least opens and shows the Qt widgets. I'm not a huge fan of the low level clear/draw calls in a higher level "visual" demo, but maybe it should still work.

**Comment by kmuehlbauer** [comment]
@djhoese Your patch works fine for me too. While tracing the qt-docs I found [this](https://doc.qt.io/qt-5/qstatusbar.html#permanent-message):

_Normal and Permanent messages are displayed by creating a small widget (QLabel, QProgressBar or even QToolButton) and then adding it to the status bar using the addWidget() or the addPermanentWidget() function. Use the removeWidget() function to remove such messages from the status bar._

I tested this with the `self.status` on the main window and it worked like your implementation within the widget.

Basically I'm almost sure that using "showMessage" doesn't work because of roundtripping event queue and using it in this example might be bad. We should just change to the above advertised method using widgets in the statusbar.

**Comment by kmuehlbauer** [comment]
> Oh and on the wiggly bar: so far if I comment out `on_draw` then it at least opens and shows the Qt widgets.

Same here, but no idea so far what to do and where to look.

**Comment by djhoese** [comment]
@kmuehlbauer For the status bar, when you do either the extra widget (`addWidget`) or my patch, do you see the same flickering effect between a small mesh then a full size mesh?

**Comment by kmuehlbauer** [comment]
@djhoese What  do you mean by small mesh vs. full size mesh.?

**Comment by djhoese** [comment]
> This "works" but it causes a weird drawing effect where the mesh goes back and forth from the right size in the center of the canvas to a smaller size in the lower left. This may be a problem related to me having a HiDPI display (which makes me a little scared).

The mesh is drawn correctly for one or two frames, then it is shown as half the size and in the lower left, then back to proper size in the center, and so on.

**Comment by kmuehlbauer** [comment]
OK, no, I do not experience these issues.
